### PR TITLE
[#12048] Migrate Reset Account Action

### PR DESCRIPTION
**Pull request by domlimm**
Part of #12048.

**Outline of Solution**

As per title.

### CODE REVIEW

File: src/main/java/teammates/storage/sqlapi/UsersDb.java [hunk 1]
```diff
@@ -277,4 +280,46 @@ public List<Student> getAllStudentsForEmail(String email) {
...
+
+        Instructor instructor = getInstructorForEmail(courseId, userEmail);
+
+        if (instructor == null) {
+            // What is a better message?
+            throw new EntityDoesNotExistException(ERROR_UPDATE_NON_EXISTENT +
+                    "Instructor [courseId=" + courseId + ", email=" + userEmail + "]");
+        }
+
+        instructor.setAccount(null);
```
Comment (samuelfangjw) [inline src/main/java/teammates/storage/sqlapi/UsersDb.java:299]: Logic should be in logic layer. Simply get instructor, instructor.setAccount(null) then cascade the update if necessary. No need to call merge anymore.

File: src/main/java/teammates/ui/webapi/ResetAccountAction.java [hunk 2]
```diff
@@ -26,11 +27,18 @@ public JsonResult execute() {
...
             }
+
+            AccountAttributes accountInfo = logic.getAccount(existingStudent.getGoogleId());
+
             wrongGoogleId = existingStudent.getGoogleId();
 
             try {
-                logic.resetStudentGoogleId(studentEmail, courseId);
-                taskQueuer.scheduleCourseRegistrationInviteToStudent(courseId, studentEmail, true);
+                if (accountInfo == null || accountInfo.isMigrated()) {
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/ResetAccountAction.java:36]:
```suggestion
                if (isAccountMigrated(googleId)) {
```

File: src/main/java/teammates/ui/webapi/ResetAccountAction.java [hunk 3]
```diff
@@ -39,16 +47,24 @@ public JsonResult execute() {
...
+                }
             } catch (EntityDoesNotExistException e) {
                 throw new EntityNotFoundException(e);
             }
         }
 
+        // To modify
         if (wrongGoogleId != null
                 && logic.getStudentsForGoogleId(wrongGoogleId).isEmpty()
                 && logic.getInstructorsForGoogleId(wrongGoogleId).isEmpty()) {
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/ResetAccountAction.java:70]: can move this into logic layer as well

File: src/main/java/teammates/storage/sqlapi/UsersDb.java (hunk 1 above)
Comment (domlimm) [reply src/main/java/teammates/storage/sqlapi/UsersDb.java]: Samuel, I did up some reading. As said before, Account is the parent of User (be it Instructor/Student).

User and Account only have a unidirectional relationship, ie. User has an attribute accountId which tells us which Account this User to referring to.

I was thinking of using any of the `Cascade` annotation maybe `C.ALL` or `C.MERGE` (not exactly sure..). All the examples I have seen so far uses Cascade from parent to child and some of which are bidirectional, but we are doing child to parent.

Not exactly sure how to cascade the update or what do you mean by `if necessary`

File: src/main/java/teammates/sqllogic/core/UsersLogic.java [hunk 4]
```diff
@@ -205,6 +212,74 @@ public List<User> getAllUsersByGoogleId(String googleId) {
...
+
+        instructor.setAccount(null);
+
+        List<Account> accounts = accountsLogic.getAccountsForEmail(email);
+
+        for (Account account : accounts) {
+            accountsLogic.deleteAccount(account.getGoogleId());
+        }
+
+        deleteAccountCascadeAll(googleId);
```
Comment (samuelfangjw) [inline src/main/java/teammates/sqllogic/core/UsersLogic.java:252]: This is not correct. Many accounts are being deleted here. We only want to delete one account if it has been orphaned. Please take the time to understand the original behaviour of the action before performing the migration.

File: src/main/java/teammates/sqllogic/core/UsersLogic.java [hunk 5]
```diff
@@ -205,6 +212,74 @@ public List<User> getAllUsersByGoogleId(String googleId) {
...
+    public void resetInstructorGoogleId(String email, String courseId, String googleId)
+            throws EntityDoesNotExistException {
+        assert email != null;
+        assert courseId != null;
+        assert googleId != null;
+
+        Instructor instructor = getInstructorForEmail(courseId, email);
+
+        if (instructor == null) {
+            // Is there a better message?
```
Comment (samuelfangjw) [inline src/main/java/teammates/sqllogic/core/UsersLogic.java:239]: Questions should not be in production code. Ugly and gives a bad impression, please remove.

File: src/test/java/teammates/sqllogic/core/AccountsLogicTest.java [hunk 6]
```diff
@@ -56,19 +57,22 @@ public void setUpMethod() {
...
 
     @Test
-    public void testDeleteAccount_accountExists_success() {
+    public void testDeleteAccount_accountExists_success()
+            throws InvalidParametersException, EntityAlreadyExistsException {
         Account account = generateTypicalAccount();
         String googleId = account.getGoogleId();
 
-        when(accountsLogic.getAccountForGoogleId(googleId)).thenReturn(account);
+        when(accountsLogic.createAccount(account)).thenReturn(account);
```
Comment (samuelfangjw) [inline src/test/java/teammates/sqllogic/core/AccountsLogicTest.java:65]: Why does a unit test for deleteAccount require createAccount to be stubbed?

File: src/main/java/teammates/ui/webapi/ResetAccountAction.java [hunk 7]
```diff
@@ -20,17 +20,19 @@ public JsonResult execute() {
...
-            wrongGoogleId = existingStudent.getGoogleId();
+
+            String existingGoogleId = existingStudent.getGoogleId();
 
             try {
-                logic.resetStudentGoogleId(studentEmail, courseId);
-                taskQueuer.scheduleCourseRegistrationInviteToStudent(courseId, studentEmail, true);
+                if (isAccountMigrated(existingGoogleId)) {
+                    sqlLogic.resetStudentGoogleId(studentEmail, courseId, existingGoogleId);
+                }
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/ResetAccountAction.java:35]: What if the account has not been migrated? This needs to work with both DBs. We need to maintain the original functionality as well, as was done in your previous PRs.

File: src/main/java/teammates/sqllogic/core/UsersLogic.java [hunk 8]
```diff
@@ -205,6 +212,74 @@ public List<User> getAllUsersByGoogleId(String googleId) {
...
+
+        student.setAccount(null);
+
+        List<Account> accounts = accountsLogic.getAccountsForEmail(email);
+
+        for (Account account : accounts) {
+            accountsLogic.deleteAccount(account.getGoogleId());
+        }
+
+        deleteAccountCascadeAll(googleId);
```
Comment (samuelfangjw) [inline src/main/java/teammates/sqllogic/core/UsersLogic.java:280]:
```suggestion
                if (usersDb.getAllUsersByGoogleId(googleId).isEmpty()) {
             accountsLogic.deleteAccountCascade(googleId);
         }
```

Please change accordingly for both.

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/sqllogic/core/UsersLogic.java (hunk 5 above)
Comment (domlimm) [reply src/main/java/teammates/sqllogic/core/UsersLogic.java:239]: Sorry, this question was meant for you during review. Will remove the comments 👍🏻

File: src/main/java/teammates/ui/webapi/ResetAccountAction.java (hunk 7 above)
Comment (domlimm) [reply src/main/java/teammates/ui/webapi/ResetAccountAction.java:35]: Sorry, I interpreted wrongly from your previous review. Will make the changes!

File: src/main/java/teammates/sqllogic/core/UsersLogic.java (hunk 4 above)
Comment (domlimm) [reply src/main/java/teammates/sqllogic/core/UsersLogic.java:252]: Got it!

File: src/test/java/teammates/sqllogic/core/AccountsLogicTest.java (hunk 6 above)
Comment (domlimm) [reply src/test/java/teammates/sqllogic/core/AccountsLogicTest.java:65]: Oops! My mistake, sorry about that!!

File: src/test/java/teammates/sqllogic/core/AccountsLogicTest.java [hunk 9]
```diff
@@ -56,19 +57,21 @@ public void setUpMethod() {
...
     @Test
-    public void testDeleteAccount_accountExists_success() {
+    public void testDeleteAccount_accountExists_success()
+            throws InvalidParametersException, EntityAlreadyExistsException {
         Account account = generateTypicalAccount();
         String googleId = account.getGoogleId();
 
-        when(accountsLogic.getAccountForGoogleId(googleId)).thenReturn(account);
-
+        accountsLogic.createAccount(account);
```
Comment (samuelfangjw) [inline src/test/java/teammates/sqllogic/core/AccountsLogicTest.java:65]: Why is this method called at all? This is a unit test for delete. We do not need to create anything.

Comment (samuelfangjw) [reply src/test/java/teammates/sqllogic/core/AccountsLogicTest.java:65]: In fact, accountsLogic.getAccountForGoogleId(googleId) will return null here. createAccount is similarly a no-op.

Comment (domlimm) [reply src/test/java/teammates/sqllogic/core/AccountsLogicTest.java:65]: > Why is this method called at all? This is a unit test for delete. We do not need to create anything.

I thought we'd need an account to delete. I have made the change and I think we are back to the existing TC from DeleteAccountAction.

> In fact, accountsLogic.getAccountForGoogleId(googleId) will return null here. createAccount is similarly a no-op.

You're right

File: src/test/java/teammates/sqllogic/core/AccountsLogicTest.java [hunk 10]
```diff
@@ -68,7 +70,8 @@ public void testDeleteAccount_accountExists_success() {
     }
 
     @Test
-    public void testDeleteAccountCascade_googleIdExists_success() {
+    public void testDeleteAccountCascade_googleIdExists_success()
+            throws InvalidParametersException, EntityAlreadyExistsException {
```
Comment (samuelfangjw) [inline src/test/java/teammates/sqllogic/core/AccountsLogicTest.java:74]: Didn't need this previously, seems like an unnecessary change to me.

File: src/main/java/teammates/ui/webapi/ResetAccountAction.java [hunk 11]
```diff
@@ -39,17 +45,23 @@ public JsonResult execute() {
...
             wrongGoogleId = existingInstructor.getGoogleId();
 
             try {
-                logic.resetInstructorGoogleId(instructorEmail, courseId);
-                taskQueuer.scheduleCourseRegistrationInviteToInstructor(null, instructorEmail, courseId, true);
+                if (isAccountMigrated(wrongGoogleId)) {
+                    sqlLogic.resetInstructorGoogleId(instructorEmail, courseId, wrongGoogleId);
+                } else {
+                    logic.resetInstructorGoogleId(instructorEmail, courseId);
+                    taskQueuer.scheduleCourseRegistrationInviteToInstructor(null, instructorEmail, courseId, true);
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/ResetAccountAction.java:56]: Where is this called for migrated entities?

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/sqllogic/core/AccountsLogicTest.java (hunk 10 above)
Comment (samuelfangjw) [reply src/test/java/teammates/sqllogic/core/AccountsLogicTest.java:74]: Likely means lint not working properly for this files. We need to fix this in another PR.

File: src/test/java/teammates/sqllogic/core/UsersLogicTest.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+        String googleId = student.getAccount().getGoogleId();
+
+        when(usersLogic.getStudentForEmail(courseId, email)).thenReturn(student);
+        when(usersDb.getAllUsersByGoogleId(googleId)).thenReturn(Collections.emptyList());
+        when(accountsLogic.getAccountForGoogleId(googleId)).thenReturn(account);
+
+        usersLogic.resetStudentGoogleId(email, courseId, googleId);
+        
+        assertEquals(null, student.getAccount());
```
Comment (samuelfangjw) [inline src/test/java/teammates/sqllogic/core/UsersLogicTest.java:108]: use assertNull instead

File: src/test/java/teammates/sqllogic/core/UsersLogicTest.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+        assertEquals(ERROR_UPDATE_NON_EXISTENT
+                + "Instructor [courseId=" + courseId + ", email=" + email + "]", exception.getMessage());
+    }
+
+    @Test
+    public void testResetStudentGoogleId_entityExistsWithNoUserAssociatedWithGoogleId_success()
+            throws EntityDoesNotExistException {
+        Account account = generateTypicalAccount();
+        Student student = getTypicalStudent();
```
Comment (samuelfangjw) [inline src/test/java/teammates/sqllogic/core/UsersLogicTest.java:94]: These should be a method field since it is used everywhere. Unit tests should ideally have as little setup as possible.

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/it/java/teammates/it/sqllogic/core/UsersLogicIT.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+import teammates.sqllogic.core.UsersLogic;
+import teammates.storage.sqlentity.Account;
+import teammates.storage.sqlentity.Course;
+import teammates.storage.sqlentity.Instructor;
+import teammates.storage.sqlentity.Student;
+
+/**
+ * SUT: {@link UsersLogic}.
+ */
+public class UsersLogicIT extends BaseTestCaseWithSqlDatabaseAccess {
```
Comment (samuelfangjw) [inline src/it/java/teammates/it/sqllogic/core/UsersLogicIT.java:24]: The test I really need to see here is the one that shows that deleting orphaned accounts works as expected. We cannot afford for this to have bugs.

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/ui/webapi/ResetAccountAction.java [hunk 15]
```diff
@@ -21,35 +23,63 @@ public JsonResult execute() {
...
         String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
         String wrongGoogleId = null;
+
         if (studentEmail != null) {
-            StudentAttributes existingStudent = logic.getStudentForEmail(courseId, studentEmail);
-            if (existingStudent == null) {
-                throw new EntityNotFoundException("Student does not exist.");
+            Student student = sqlLogic.getStudentForEmail(courseId, studentEmail);
+
+            // To handle check for migration for IT. To remove once migrated.
```
[the remark this answers is not in the document]
Comment (domlimm) [reply src/main/java/teammates/ui/webapi/ResetAccountAction.java:30]: Have removed it with your comment on the wrong logic 👍🏻 Sorry for the mistake..

**Review by samuelfangjw — APPROVED** [review]
